source/extensions/quic_listeners/quiche/envoy_quic_client_session.cc

```diff
@@ -48,12 +48,12 @@ void EnvoyQuicClientSession::Initialize() {
 }
 
 void EnvoyQuicClientSession::OnCanWrite() {
-  const uint64_t headers_to_send_old =
-      quic::VersionUsesHttp3(transport_version()) ? 0u : headers_stream()->BufferedDataBytes();
-  quic::QuicSpdyClientSession::OnCanWrite();
-  const uint64_t headers_to_send_new =
-      quic::VersionUsesHttp3(transport_version()) ? 0u : headers_stream()->BufferedDataBytes();
-  adjustBytesToSend(headers_to_send_new - headers_to_send_old);
+  if (quic::VersionUsesHttp3(transport_version())) {
+    quic::QuicSpdyClientSession::OnCanWrite();
+  } else {
+    SendBufferMonitor::ScopedWatermarkBufferUpdater updater(headers_stream(), this);
```

Comment:
mind commenting inline why we need watermarks here and not above?

Also just to make sure, do we know of endpoints which only support gQUIC?  I assume it's too early to only support IETF quic and simplify our implementations, but I figure it's worth checking.  cc @DavidSchinazi

Reply:
Today servers behind GCLB support gQUIC but not IETF QUIC. We're expecting to get them IETF QUIC support in Q2 2021.

The distinction is that in IETF QUIC we send the headers on the request stream, whereas gQUIC sends them in the headers stream, and I agree that a comment to this effect would be helpful.